File: Exercism/haskell/twelve-days/src/TwelveDays.hs
module TwelveDays (recite) where

import Data.List (intercalate)
import Text.Printf (printf)


newtype SInt = SInt { getInt::Int } deriving (Eq, Ord)

instance Show SInt where
  show i = case getInt i of
             1 -> "one"
             2 -> "two"
             3 -> "three"
             4 -> "four"
             5 -> "five"
             6 -> "six"
             7 -> "seven"
             8 -> "eight"
             9 -> "nine"
             10 -> "ten"
             11 -> "eleven"
             12 -> "twelve"
             _  -> error "I do not know how to represent " ++ show i

nth :: Int -> String
nth n = case n of
          1  -> "first"
          2  -> "second"
          3  -> "third"
          5  -> "fifth"
          8  -> "eighth"
          9  -> "ninth"
          12 -> "twelfth"
          _  -> show (SInt n) ++ "th"

-- Right, stop it! Started off with a nice little idea about grannies attacking young men, but now it's got silly.
-- (That is, it started out with 1,2,3 being exceptions but then look! more exceptions than the rule!)
-- (Oh, well.)

gifts :: [String]
gifts = reverse [ "Drummers Drumming"
                , "Pipers Piping"
                , "Lords-a-Leaping"
                , "Ladies Dancing"
                , "Maids-a-Milking"
                , "Swans-a-Swimming"
                , "Geese-a-Laying"
                , "Gold Rings"
                , "Calling Birds"
                , "French Hens"
                , "Turtle Doves"
                , "Partridge in a Pear Tree"
                ]

recite :: Int -> Int -> [String]
recite start stop = [stanza n | n <- [start..stop]] where
  fmt = "On the %s day of Christmas my true love gave to me: %s."
  stanza n = printf fmt (nth n) (intercalate ", " giftsOfStanza) where
    giftsOfStanza = reverse $ zipWith spellCount [1..n] (take n gifts)
    spellCount i = unwords . (count i :) . (:[])
    count i
      | i == 1 = if n == 1 then "a" else "and a"
      | otherwise = show (SInt i)

File: Exercism/python/twelve-days/twelve_days.py
SPELL_NUMBER = [
    ("a", "first"),
    ("two", "second"),
    ("three", "third"),
    ("four", "fourth"),
    ("five", "fifth"),
    ("six", "sixth"),
    ("seven", "seventh"),
    ("eight", "eighth"),
    ("nine", "ninth"),
    ("ten", "tenth"),
    ("eleven", "eleventh"),
    ("twelve", "twelfth"),
]


def spell(n: int) -> str:
    return SPELL_NUMBER[n - 1][0]


def nth(n: int) -> str:
    return SPELL_NUMBER[n - 1][1]


GIFTS = [
    "Partridge in a Pear Tree",
    "Turtle Doves",
    "French Hens",
    "Calling Birds",
    "Gold Rings",
    "Geese-a-Laying",
    "Swans-a-Swimming",
    "Maids-a-Milking",
    "Ladies Dancing",
    "Lords-a-Leaping",
    "Pipers Piping",
    "Drummers Drumming",
]


def gifts(n: int) -> str:
    and_ = "and " if n > 1 else ""
    return ", ".join(
        reversed([
            ("" if i > 1 else and_) + spell(i) + " " + gift
            for i, gift in enumerate(GIFTS[:n], 1)
        ])
    )


def recite(start_verse: int, end_verse: int) -> [str]:
    return [
        f"On the {nth(n)} day of Christmas my true love gave to me: {gifts(n)}."
        for n in range(start_verse, end_verse + 1)
    ]
